File: OpenMeteo.cs
using System.Globalization;
using System.Net.Http;
using System.Text.Json;

namespace WorldClock;

public interface IWeatherService
{
    /// <summary>Current weather for each location, in the same order. Throws on network or API failure.</summary>
    Task<IReadOnlyList<CurrentWeather>> GetCurrentAsync(IReadOnlyList<(double Lat, double Lon)> locations, CancellationToken ct = default);
}

public interface IPlaceSearch
{
    /// <summary>Places matching the name. Throws on network or API failure.</summary>
    Task<IReadOnlyList<PlaceResult>> SearchAsync(string name, CancellationToken ct = default);
}

/// <summary>
/// Weather and place search from open-meteo.com. Free, no API key, non-commercial use only, CC BY 4.0.
/// </summary>
public sealed class OpenMeteoClient : IWeatherService, IPlaceSearch
{
    public static OpenMeteoClient Shared { get; } = new();

    private static readonly HttpClient Http = CreateHttpClient();

    private static HttpClient CreateHttpClient()
    {
        var http = new HttpClient { Timeout = TimeSpan.FromSeconds(10) };
        http.DefaultRequestHeaders.UserAgent.ParseAdd($"WorldClock/{AppInfo.Version} (+https://github.com/stewartisland/world-clock)");
        return http;
    }

    public async Task<IReadOnlyList<CurrentWeather>> GetCurrentAsync(
        IReadOnlyList<(double Lat, double Lon)> locations, CancellationToken ct = default)
    {
        if (locations.Count == 0) return [];

        var lats = string.Join(',', locations.Select(l => Format(l.Lat)));
        var lons = string.Join(',', locations.Select(l => Format(l.Lon)));
        var url = $"https://api.open-meteo.com/v1/forecast?latitude={lats}&longitude={lons}" +
                  "&current=temperature_2m,weather_code,is_day&temperature_unit=celsius";

        using var doc = JsonDocument.Parse(await Http.GetStringAsync(url, ct));

        // One location returns an object; several return an array in request order.
        var items = doc.RootElement.ValueKind == JsonValueKind.Array
            ? doc.RootElement.EnumerateArray().ToList()
            : [doc.RootElement];

        return items.Select(item =>
        {
            var current = item.GetProperty("current");
            return new CurrentWeather(
                current.GetProperty("temperature_2m").GetDouble(),
                current.GetProperty("weather_code").GetInt32(),
                current.GetProperty("is_day").GetInt32() == 1);
        }).ToList();
    }

    public async Task<IReadOnlyList<PlaceResult>> SearchAsync(string name, CancellationToken ct = default)
    {
        var url = "https://geocoding-api.open-meteo.com/v1/search?count=10&language=en&format=json&name=" +
                  Uri.EscapeDataString(name);

        using var doc = JsonDocument.Parse(await Http.GetStringAsync(url, ct));
        if (!doc.RootElement.TryGetProperty("results", out var results))
            return [];

        return results.EnumerateArray().Select(r =>
        {
            var placeName = GetString(r, "name") ?? "";
            var parts = new[] { placeName, GetString(r, "admin1"), GetString(r, "country") }
                .Where(p => !string.IsNullOrEmpty(p))
                .Distinct();
            var iana = GetString(r, "timezone");
            string? windowsId = iana is not null && TimeZoneInfo.TryConvertIanaIdToWindowsId(iana, out var id) ? id : null;

            return new PlaceResult(
                placeName,
                string.Join(", ", parts),
                Math.Round(r.GetProperty("latitude").GetDouble(), 4),
                Math.Round(r.GetProperty("longitude").GetDouble(), 4),
                windowsId);
        }).ToList();
    }

    private static string? GetString(JsonElement e, string name) =>
        e.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;

    private static string Format(double d) => d.ToString("0.####", CultureInfo.InvariantCulture);
}

File: ios/WorldClock/OpenMeteo.swift
import Foundation

protocol WeatherService: Sendable {
    /// Current weather for each location, in the same order. Throws on network or API failure.
    func current(for locations: [(lat: Double, lon: Double)]) async throws -> [CurrentWeather]
}

protocol PlaceSearch: Sendable {
    /// Places matching the name. Throws on network or API failure.
    func search(_ name: String) async throws -> [PlaceResult]
}

/// Weather and place search from open-meteo.com. Free, no API key, non-commercial use only, CC BY 4.0.
struct OpenMeteoClient: WeatherService, PlaceSearch {
    static let shared = OpenMeteoClient()

    private let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 10
        config.httpAdditionalHeaders = [
            "User-Agent": "WorldClock-iOS/\(AppInfo.version) (+https://github.com/stewartisland/world-clock)",
        ]
        return URLSession(configuration: config)
    }()

    func current(for locations: [(lat: Double, lon: Double)]) async throws -> [CurrentWeather] {
        if locations.isEmpty { return [] }

        let lats = locations.map { format($0.lat) }.joined(separator: ",")
        let lons = locations.map { format($0.lon) }.joined(separator: ",")
        let url = URL(string: "https://api.open-meteo.com/v1/forecast?latitude=\(lats)&longitude=\(lons)"
            + "&current=temperature_2m,weather_code,is_day&temperature_unit=celsius")!
        return try Self.parseWeather(try await get(url))
    }

    func search(_ name: String) async throws -> [PlaceResult] {
        var components = URLComponents(string: "https://geocoding-api.open-meteo.com/v1/search")!
        components.queryItems = [
            URLQueryItem(name: "count", value: "10"),
            URLQueryItem(name: "language", value: "en"),
            URLQueryItem(name: "format", value: "json"),
            URLQueryItem(name: "name", value: name),
        ]
        return try Self.parsePlaces(try await get(components.url!))
    }

    private func get(_ url: URL) async throws -> Data {
        let (data, response) = try await session.data(from: url)
        guard (response as? HTTPURLResponse)?.statusCode == 200 else { throw URLError(.badServerResponse) }
        return data
    }

    private func format(_ value: Double) -> String { String(format: "%.4f", value) }

    // Parsing, kept separate so it can be tested without the network.

    private struct Forecast: Decodable {
        struct Current: Decodable {
            let temperature_2m: Double
            let weather_code: Int
            let is_day: Int
        }
        let current: Current
    }

    static func parseWeather(_ data: Data) throws -> [CurrentWeather] {
        // One location returns an object; several return an array in request order.
        let decoder = JSONDecoder()
        let items = try (try? decoder.decode([Forecast].self, from: data)) ?? [decoder.decode(Forecast.self, from: data)]
        return items.map {
            CurrentWeather(temperatureC: $0.current.temperature_2m, weatherCode: $0.current.weather_code, isDay: $0.current.is_day == 1)
        }
    }

    private struct GeocodingResponse: Decodable {
        struct Place: Decodable {
            let name: String?
            let admin1: String?
            let country: String?
            let latitude: Double
            let longitude: Double
            let timezone: String?
        }
        let results: [Place]?
    }

    static func parsePlaces(_ data: Data) throws -> [PlaceResult] {
        let response = try JSONDecoder().decode(GeocodingResponse.self, from: data)
        return (response.results ?? []).map { place in
            let name = place.name ?? ""
            var parts: [String] = []
            for part in [name, place.admin1, place.country].compactMap({ $0 }) where !part.isEmpty && !parts.contains(part) {
                parts.append(part)
            }
            let zone = place.timezone.flatMap { TimeZone(identifier: $0) }?.identifier
            return PlaceResult(
                name: name,
                displayName: parts.joined(separator: ", "),
                lat: (place.latitude * 10_000).rounded() / 10_000,
                lon: (place.longitude * 10_000).rounded() / 10_000,
                timeZoneId: zone)
        }
    }
}
